File: tests/consistency/schema_contract_test.py
"""Phia Python cua hop dong schema.

Ban Node doi xung o tests/consistency/schema_contract.test.mjs, CHAY CUNG MOT FILE
FIXTURE (charts/fixtures/schema-cases.json). Khi mot ben doi luat ma ben kia quen, mot
trong hai bo test do ngay.

Doi chieu bang MA LOI chu khong bang cau chu tieng Viet.
"""

import json
import sys
from pathlib import Path

import pytest

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "charts" / "matplotlib"))

from schema import (  # noqa: E402
    ERR,
    LoiSchema,
    SCHEMA_VERSION,
    VOCAB,
    cach_ve,
    co_co,
    so_thap_phan,
    validate_period,
    validate_row,
    validate_series,
)

with (ROOT / "charts" / "fixtures" / "schema-cases.json").open(encoding="utf-8") as f:
    FIXTURE = json.load(f)

CHAY = {
    "series": validate_series,
    "row": lambda d: validate_row(d, 0),
    "period": validate_period,
}


def test_moi_ca_trong_fixture_vang_cho_dung_ket_qua():
    hong = []
    for ca in FIXTURE["cases"]:
        chay = CHAY.get(ca["loai"])
        assert chay is not None, f'ca "{ca["ten"]}" co loai "{ca["loai"]}" khong ai xu ly'
        if ca["mong_doi"] == "pass":
            try:
                chay(ca["du_lieu"])
            except LoiSchema as e:
                hong.append(f'"{ca["ten"]}": mong doi PASS nhung nem {e.ma}')
        else:
            da_nem = None
            try:
                chay(ca["du_lieu"])
            except LoiSchema as e:
                da_nem = e.ma
            if da_nem is None:
                hong.append(f'"{ca["ten"]}": mong doi FAIL voi {ca["ma_loi"]} nhung PASS')
            elif da_nem != ca["ma_loi"]:
                hong.append(f'"{ca["ten"]}": mong doi {ca["ma_loi"]} nhung nem {da_nem}')
    assert not hong, f"{len(hong)} ca lech hop dong:\n" + "\n".join(hong)


def test_bo_ca_phu_het_ma_loi_da_khai():
    ma_duoc_kiem = {c["ma_loi"] for c in FIXTURE["cases"] if c.get("ma_loi")}
    # Hai ma nay khong den tu du lieu dau vao ma tu loi lap trinh trong preset.
    mien_tru = {"VALUE_SAI_KIEU", "CHI_SO_NGOAI_PHAM_VI"}
    tat_ca = {v for k, v in vars(ERR).items() if not k.startswith("_") and isinstance(v, str)}
    thieu = sorted(tat_ca - ma_duoc_kiem - mien_tru)
    assert not thieu, (
        f"cac ma loi sau chua co ca fixture nao kiem: {', '.join(thieu)}. "
        "Them ca vao charts/fixtures/schema-cases.json"
    )


def test_ma_loi_hai_ban_trung_khop():
    """Kiem doi xung: moi ma ben Python phai co ben Node va nguoc lai.

    Doc thang file schema.mjs bang van ban thay vi chay Node, de test nay khong phu
    thuoc vao viec co Node trong moi truong chay pytest.
    """
    mjs = (ROOT / "charts" / "echarts" / "schema.mjs").read_text(encoding="utf-8")
    py_ma = {v for k, v in vars(ERR).items() if not k.startswith("_") and isinstance(v, str)}
    thieu_ben_node = sorted(m for m in py_ma if f"{m}: '{m}'" not in mjs)
    assert not thieu_ben_node, (
        f"ma loi co ben Python nhung khong thay ben Node: {', '.join(thieu_ben_node)}"
    )


def test_so_thap_phan_lay_tu_tu_vung():
    assert so_thap_phan({"unit": "ty_dong"}) == 0
    assert so_thap_phan({"unit": "phan_tram"}) == 1
    assert so_thap_phan({"unit": "phan_tram", "decimals": 2}) == 2


def test_cach_ve_phan_biet_du_ba_loai_gia_tri_thieu():
    dau_hieu = [
        cach_ve("chua_cong_bo")["danh_dau_truc"],
        cach_ve("khong_ton_tai")["danh_dau_truc"],
        cach_ve("loai_bat_thuong")["danh_dau_truc"],
    ]
    assert len(set(map(str, dau_hieu))) == 3, (
        f"ba trang thai thieu phai co ba dau hieu khac nhau, dang la {dau_hieu}"
    )
    assert cach_ve("co_that")["noi_duong"] is True


def test_co_tinh_tu_chi_so_nguyen():
    la_base = co_co(2, 5)
    assert la_base(2) is True
    assert la_base(3) is False
    with pytest.raises(LoiSchema) as e1:
        co_co(5, 5)
    assert e1.value.ma == ERR.CHI_SO_NGOAI_PHAM_VI
    with pytest.raises(LoiSchema) as e2:
        co_co(True, 5)  # bool la subclass cua int, phai bi chan tuong minh
    assert e2.value.ma == ERR.CHI_SO_NGOAI_PHAM_VI


def test_tu_vung_va_phien_ban():
    assert SCHEMA_VERSION.count(".") == 2
    assert len(VOCAB["unit"]) >= 8
    for ten, mo in VOCAB["unit"].items():
        assert isinstance(mo["decimals"], int), f"don vi {ten} thieu decimals mac dinh"
        assert mo["nhan"], f"don vi {ten} thieu nhan hien thi"

File: tests/consistency/schema_contract.test.mjs
// Phia Node cua hop dong schema. Ban Python doi xung o
// tests/consistency/schema_contract_test.py, CHAY CUNG MOT FILE FIXTURE.
//
// Y do: phan lon thu troi dat giua hai ban schema la tu vung va luat ngu nghia. Tu
// vung da tach ra charts/schema.vocab.json nen ca hai cung doc. Luat ngu nghia van
// phai viet tay hai lan, va cach chong troi duy nhat that su hieu qua la bat ca hai
// chay cung mot bo ca, moi ca ghi ro phai pass hay fail va fail voi ma nao.
//
// Doi chieu bang MA LOI chu khong bang cau chu tieng Viet: sua loi nhan cho de hieu
// hon la viec tot, va no khong duoc lam do test o phia ben kia.
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import path from 'node:path';
import {
  validateSeries,
  validateRow,
  validatePeriod,
  SCHEMA_VERSION,
  VOCAB,
  soThapPhan,
  cachVe,
  coCo,
  ERR,
} from '../../charts/echarts/schema.mjs';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');
const FIXTURE = JSON.parse(
  readFileSync(path.join(ROOT, 'charts/fixtures/schema-cases.json'), 'utf8'),
);

const CHAY = {
  series: (d) => validateSeries(d),
  row: (d) => validateRow(d, 0),
  period: (d) => validatePeriod(d),
};

test('moi ca trong fixture vang cho dung ket qua mong doi', () => {
  const hong = [];
  for (const ca of FIXTURE.cases) {
    const chay = CHAY[ca.loai];
    assert.ok(chay, `ca "${ca.ten}" co loai "${ca.loai}" khong ai xu ly`);
    if (ca.mong_doi === 'pass') {
      try {
        chay(ca.du_lieu);
      } catch (e) {
        hong.push(`"${ca.ten}": mong doi PASS nhung nem ${e.ma || e.message}`);
      }
    } else {
      let daNem = null;
      try {
        chay(ca.du_lieu);
      } catch (e) {
        daNem = e.ma;
      }
      if (daNem === null) hong.push(`"${ca.ten}": mong doi FAIL voi ${ca.ma_loi} nhung PASS`);
      else if (daNem !== ca.ma_loi) hong.push(`"${ca.ten}": mong doi ${ca.ma_loi} nhung nem ${daNem}`);
    }
  }
  assert.deepEqual(hong, [], `${hong.length} ca lech hop dong:\n${hong.join('\n')}`);
});

test('bo ca phu het cac ma loi da khai, khong de ma nao khong duoc kiem', () => {
  const maDuocKiem = new Set(FIXTURE.cases.filter((c) => c.ma_loi).map((c) => c.ma_loi));
  // Hai ma duoi day khong den tu du lieu dau vao ma tu loi lap trinh trong preset, nen
  // khong co ca fixture: VALUE_SAI_KIEU la ma du phong, CHI_SO_NGOAI_PHAM_VI do coCo()
  // nem khi preset truyen chi so sai.
  const MIEN_TRU = new Set(['VALUE_SAI_KIEU', 'CHI_SO_NGOAI_PHAM_VI']);
  const thieu = Object.values(ERR).filter((m) => !maDuocKiem.has(m) && !MIEN_TRU.has(m));
  assert.deepEqual(
    thieu,
    [],
    `cac ma loi sau chua co ca fixture nao kiem: ${thieu.join(', ')}. Them ca vao charts/fixtures/schema-cases.json`,
  );
});

test('so chu so thap phan lay tu tu vung khi series khong khai', () => {
  assert.equal(soThapPhan({ unit: 'ty_dong' }), 0);
  assert.equal(soThapPhan({ unit: 'phan_tram' }), 1);
  assert.equal(soThapPhan({ unit: 'phan_tram', decimals: 2 }), 2, 'khai tuong minh phai thang tu vung');
});

test('cach ve phan biet du ba loai gia tri thieu', () => {
  const chuaCongBo = cachVe('chua_cong_bo');
  const khongTonTai = cachVe('khong_ton_tai');
  const daLoai = cachVe('loai_bat_thuong');
  // Ca ba deu khong ve diem, nhung phai KHAC NHAU o dau hieu tren truc, neu khong thi
  // gop chung lam mot con hon la giu ba ten goi tao cam giac phan biet ma khong phan biet.
  const dauHieu = [chuaCongBo, khongTonTai, daLoai].map((c) => c.danh_dau_truc);
  assert.equal(new Set(dauHieu).size, 3, `ba trang thai thieu phai co ba dau hieu khac nhau, dang la ${JSON.stringify(dauHieu)}`);
  assert.equal(cachVe('co_that').noi_duong, true);
});

test('co tinh tu chi so nguyen, khong so sanh so thuc', () => {
  const laBase = coCo(2, 5);
  assert.equal(laBase(2), true);
  assert.equal(laBase(3), false);
  assert.throws(() => coCo(5, 5), /CHI_SO_NGOAI_PHAM_VI/, 'chi so bang do dai mang phai bi chan');
  assert.throws(() => coCo(1.5, 5), /CHI_SO_NGOAI_PHAM_VI/, 'chi so thuc phai bi chan');
});

test('tu vung va phien ban doc duoc tu file dung chung', () => {
  assert.match(SCHEMA_VERSION, /^\d+\.\d+\.\d+$/);
  assert.ok(Object.keys(VOCAB.unit).length >= 8, 'tu vung don vi qua ngheo');
  for (const [ten, mo] of Object.entries(VOCAB.unit)) {
    assert.ok(Number.isInteger(mo.decimals), `don vi ${ten} thieu decimals mac dinh`);
    assert.ok(mo.nhan && mo.nhan.length > 0, `don vi ${ten} thieu nhan hien thi`);
  }
});
